File: List/RotateList/RotateList.java
/*
class ListNode {
	private int val;
	ListNode next;
	ListNode(int v)	{
		this.val = v;
		this.next = null
	}
}
*/
class Solution {
	public ListNode rotateList(ListNode head, int k) {
		if(head == null)
			return null;
		
		ListNode slow = head;
		ListNode fast = head;
		int len = 1;
		for(; fast.next != null && len <= k; ++len) {
			fast = fast.next;
		}
		
		if(len <= k) {
			k %= len;
			fast = head;
			for(int i = 0; i < k; i++) {
				fast = fast.next;
			}
		}
		
		while(fast != null) {
			fast = fast.next;
			slow = slow.next;
		}
		
		fast.next = head;
		head = slow.next;
		slow.next = null;
				
		return head;
	}
}

File: List/RotateList/RotateList.py
"""
class ListNode:
    def __init__(self, v):
        self.val = v
        self.head = None
"""

class Solution:
    def rotateList(head, k):
        
        if head is None:
            return
            
        slow = head
        fast = head
        len = 1
        
        for len in xrange(k) and fast is not None:
            fast = fast.next
            
        if len <= k:
            len %= k
            for i in xrange(k):
                fast = fast.next
                
        fast.next = head
        head = slow.next
        slow.next = None
        
        return head
        